File: apps/desktop/src-tauri/src/sessions.rs
use crate::error::AppResult;
use crate::paths::{get_sessions_dir, get_sessions_file};
use crate::state::{SessionMeta, SessionsStore};
use std::fs;

/// Load the sessions store from disk
fn load_sessions_store() -> SessionsStore {
    let file_path = get_sessions_file();
    if file_path.exists() {
        let content = fs::read_to_string(&file_path).unwrap_or_default();
        serde_json::from_str(&content).unwrap_or_default()
    } else {
        SessionsStore::default()
    }
}

// =============================================================================
// Tauri Commands
// =============================================================================

#[tauri::command]
pub fn create_session_dir(session_id: String) -> AppResult<String> {
    let session_dir = get_sessions_dir().join(&session_id);
    fs::create_dir_all(&session_dir)?;

    // Create screens directory by default so agent doesn't need to mkdir
    fs::create_dir_all(session_dir.join("screens"))?;

    Ok(session_dir.to_string_lossy().to_string())
}

#[tauri::command]
pub fn get_session_cwd(session_id: String) -> String {
    get_sessions_dir()
        .join(&session_id)
        .to_string_lossy()
        .to_string()
}

#[tauri::command]
pub fn save_session_metadata(session: SessionMeta) -> AppResult<()> {
    let file_path = get_sessions_file();

    if let Some(parent) = file_path.parent() {
        fs::create_dir_all(parent)?;
    }

    let mut store = load_sessions_store();

    if let Some(existing) = store.sessions.iter_mut().find(|s| s.id == session.id) {
        *existing = session;
    } else {
        store.sessions.push(session);
    }

    let json = serde_json::to_string_pretty(&store)?;
    fs::write(&file_path, json)?;

    Ok(())
}

#[tauri::command]
pub fn load_sessions_metadata() -> Vec<SessionMeta> {
    load_sessions_store().sessions
}

#[tauri::command]
pub fn delete_session_metadata(session_id: String) -> AppResult<()> {
    let file_path = get_sessions_file();
    let mut store = load_sessions_store();

    store.sessions.retain(|s| s.id != session_id);

    let json = serde_json::to_string_pretty(&store)?;
    fs::write(&file_path, json)?;

    let session_dir = get_sessions_dir().join(&session_id);
    if session_dir.exists() {
        fs::remove_dir_all(&session_dir)?;
    }

    Ok(())
}

#[tauri::command]
pub fn toggle_session_favorite(session_id: String) -> AppResult<bool> {
    let file_path = get_sessions_file();
    let mut store = load_sessions_store();

    let session = store
        .sessions
        .iter_mut()
        .find(|s| s.id == session_id)
        .ok_or_else(|| crate::error::AppError::Custom(format!("Session {} not found", session_id)))?;

    session.favorite = !session.favorite;
    let new_favorite = session.favorite;

    let json = serde_json::to_string_pretty(&store)?;
    fs::write(&file_path, json)?;

    Ok(new_favorite)
}

File: apps/desktop/electron/ipc/sessions.ts
import fs from "node:fs"
import fsp from "node:fs/promises"
import path from "node:path"
import type { SessionMeta } from "@dilag/desktop-bridge"
import { getSessionsDir, getSessionsFile } from "./paths.js"

type SessionsStore = { sessions: SessionMeta[] }

function loadSessionsStore(): SessionsStore {
  try {
    return JSON.parse(fs.readFileSync(getSessionsFile(), "utf8"))
  } catch {
    return { sessions: [] }
  }
}

async function saveSessionsStore(store: SessionsStore) {
  await fsp.mkdir(path.dirname(getSessionsFile()), { recursive: true })
  await fsp.writeFile(getSessionsFile(), JSON.stringify(store, null, 2))
}

export async function createSessionDir(sessionId: string): Promise<string> {
  const sessionDir = path.join(getSessionsDir(), sessionId)
  await fsp.mkdir(path.join(sessionDir, "screens"), { recursive: true })
  return sessionDir
}

export function getSessionCwd(sessionId?: string): string {
  return path.join(getSessionsDir(), sessionId ?? "")
}

export function loadSessionsMeta(): SessionMeta[] {
  return loadSessionsStore().sessions
}

export async function saveSessionMeta(session: SessionMeta): Promise<void> {
  const store = loadSessionsStore()
  const index = store.sessions.findIndex((item) => item.id === session.id)
  if (index >= 0) store.sessions[index] = session
  else store.sessions.push(session)
  await saveSessionsStore(store)
}

export async function deleteSessionMeta(sessionId: string): Promise<void> {
  const store = loadSessionsStore()
  store.sessions = store.sessions.filter((session) => session.id !== sessionId)
  await saveSessionsStore(store)
  await fsp.rm(path.join(getSessionsDir(), sessionId), { recursive: true, force: true })
}

export async function toggleSessionFavorite(sessionId: string): Promise<boolean> {
  const store = loadSessionsStore()
  const session = store.sessions.find((item) => item.id === sessionId)
  if (!session) throw new Error(`Session ${sessionId} not found`)
  session.favorite = !session.favorite
  await saveSessionsStore(store)
  return session.favorite
}
